### PR TITLE
Implement alphanum sorting in sortable.js

### PR DESCRIPTION
**Pull request by ssbarnea**
JENKINS-41577 alphanum sorting for views
JENKINS-44119 alphanum sorting epic

# Description

See [JENKINS-41577](https://issues.jenkins-ci.org/browse/JENKINS-41577) and [JENKINS-44119](https://issues.jenkins-ci.org/browse/JENKINS-44119)

Details: Jobs sorting is now using alphanum sorting.

### Changelog entries

Proposed changelog entries:

* Entry 1: JENKINS-41577, Sorting is now user alphanum algorithm

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Link to JIRA ticket in description, if appropriate
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc)

### Desired reviewers

@oleg-nenashev @daniel-beck

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
I am definitely a wrong guy to review the JS code. Maybe @recena?

**Comment by daniel-beck** [comment]
IMO needs more manual testing than review.

File: war/src/main/webapp/scripts/sortable.js [hunk 1]
```diff
@@ -266,6 +270,35 @@ var Sortable = (function() {
...
+        for (x = 0; aa[x] && bb[x]; x++) {
+          if (aa[x] !== bb[x]) {
+            var c = Number(aa[x]), d = Number(bb[x]);
+            if (c == aa[x] && d == bb[x]) {
+              return c - d;
+            } else return (aa[x] > bb[x]) ? 1 : -1;
+          }
+        }
+        return aa.length - bb.length;
+    }
```
Comment (orrc) [inline war/src/main/webapp/scripts/sortable.js]: This code seems to be copied from elsewhere. Do you know what the license is?

Comment (ssbarnea) [inline war/src/main/webapp/scripts/sortable.js]: Yes, that code is mainly based on the http://www.davekoelle.com/files/alphanum.js which is LGPL. Based on https://softwareengineering.stackexchange.com/questions/141847/how-does-using-a-lgpl-gem-affect-my-mit-licensed-application i would assume that there is no license conflict.

Please let me know if you want to make any changes to the code as we would all benefit from merging this.

The code worked for me well in Firefox, Chrome and Safari. I didn't had any other browsers to test with but based on its nature I doubt we would encounter any browser specific issues.

Comment (daniel-beck) [inline war/src/main/webapp/scripts/sortable.js]: > Only if you'd modify the library or directly include parts of the code in your software, then your code would have to be LGPL.

This seems to pretty obviously taint Jenkins.

**Review by daniel-beck — CHANGES_REQUESTED** [review]
License issues

File: war/src/main/webapp/scripts/sortable.js (hunk 1 above)
Comment (dmkoelle) [inline war/src/main/webapp/scripts/sortable.js]: I'm the author of the Alphanum algorithm and I have no specific reason to maintain the LGPL license on it. I'd like to change Alphanum's licensing and I might as well select a license that will satisfy the use of the code in Jenkins. I imagine an MIT license would fit nicely with Jenkins. Would an Apache license also be acceptable?

**Comment by ssbarnea** [comment]
@daniel-beck I contacted the author and today I got a reply as he would be glad to contribute these lines of code to Jenkins. You should hear back from him here for confirmation. His pick of LGPL was a moment decision and not something designed to harden the adoption by other projects.

**Comment by dmkoelle** [comment]
(I also commented above the the thread started by @orrc)

I'm the author of the Alphanum algorithm I'm interested in licenses other than LGPL that would make this kind of use of Alphanum as easy as possible. You have my permission to proceed.

**Comment by oleg-nenashev** [comment]
@daniel-beck @orrc So WDYT?

**Comment by daniel-beck** [comment]
IANAL and all that, but I still see 2-3 problems here:

* @dmkoelle 's statement above does not change the license;
* the JS adaptation was independently developed and appears to be LGPL only (unclear license header);
* And finally, I'd prefer if this were integrated as a separate, as-unmodified-as-possible source file, with attribution/copyright and original license included.

 ideally @dmkoelle publishes his original work with MIT license, gets Brian Husman to re-release `alphanum.js` explicitly as MIT (or @ssbarnea wipes his memory clean of this implementation and creates his own based solely on @dmkoelle 's work, presumably the Java version), and @ssbarnea adapts the PR to integrate a new `alphanum.js`

**Comment by daniel-beck** [comment]
Relevant part of the governance document (bold highlights mine): https://jenkins.io/project/governance/#copying-code-from-elsewhere

> **When you have a license to do so, and when that license is compatible with the MIT license, you can copy the code from elsewhere into Jenkins.**
> 
> The most typical case of this is that the original code is licensed under a certain subset of the open-source licenses, such as ASL, BSD, and MIT license. **Copyleft licenses, even though they are open-sourced, cannot be copied, such as EPL and GPL.**
> 
> **The code to be copied must be clearly marked with the license it is under, and when copying, you need to maintain the copyright/license attribution in the header.** Please also indicate the origin of the copy as a part of the commit message.

**Comment by dmkoelle** [comment]
I will be changing alphanum to the MIT License as soon as possible (next day or two), and will update this thread when I complete that work.

**Comment by johnou** [comment]
@dmkoelle any updates?

**Comment by daniel-beck** [comment]
I tried using the JS implementation of this for https://github.com/jenkins-infra/infra-statistics/pull/32 before choosing a much simpler approach, and noticed that it sorts  `2.10`, `2.2`, `2.7`. Does this also apply to this implementation?

**Comment by dmkoelle** [comment]
Great news - the Alphanum algorithm now officially uses the MIT License! 

Please see http://davekoelle.com/alphanum.html

Additionally, I have updated the license text in each of the source files linked from http://davekoelle.com/alphanum.html

**Comment by dmkoelle** [comment]
@daniel-beck I have verified that the implementation as I had written it would sort those values as 2.2, 2.7, 2.10.

**Comment by ssbarnea** [comment]
Please let me know if further changes are expected from me on this PR. If so please tell me so I can make them, merge and move on to more pressing tasks ;)

**Comment by johnou** [comment]
@ssbarnea I would think taking a fresh copy of the re-licensed code, keeping the licence headers in place and updating the pull request would be a good start :)

**Comment by daniel-beck** [comment]
@ssbarnea See https://github.com/jenkinsci/jenkins/pull/2885#issuecomment-309266874

**Comment by daniel-beck** [comment]
@dmkoelle Did you get permission from Brian Huisman, author of the JS implementation, to do so, or otherwise have an indication that it would be licensed under whatever license the original is? All we have is the header to go on, and while it states "same as original", it also includes (included) the LGPL.

**Comment by daniel-beck** [comment]
@ssbarnea What is this PR supposed to do again? Or is my observation above, with the sort implementation being broken, correct? I tried a PR build, and got this on the "All" view for projects (no display name, these are their actual names):

> screen shot

I clicked the column header a few times, but it never showed up in the expected order, either ascending or descending.

**Comment by dmkoelle** [comment]
@daniel-beck Brian Huisman gives his permission. In an email to me on July 12, 2017, he wrote: "I have no problem with changing the license."

**Comment by daniel-beck** [comment]
@dmkoelle Great, thank you. So from a license POV I have no concerns anymore.

**Comment by johnou** [comment]
@daniel-beck wouldn't it still require updating this pull request with the updated code and license headers in tact? I have seen other projects revert / downgrade libraries because the license changed and it was no longer compatible / tainted the project.

**Comment by daniel-beck** [comment]
@johnou Right, to clarify what I mean:

* [x] License is compatible now and no longer an obstacle to inclusion.
* [ ] Needs updating the implementation (separate file, license header, etc.) as previously mentioned
* [ ] Needs a response to my concerns about possible buggy implementation

**Comment by dmkoelle** [comment]
I suspect, but have not confirmed, that the issue is with the line "var m = (i == 46 || (i >=48 && i <= 57));" in the code - specifically, the check for "i == 46", where ASCII 46 is a period. That is not something that my original implementation checks for (I check for whether the character is purely a digit), and I think this is causing 2.10 to come before 2.3 because, mathematically, "2 point 1 0" does indeed come before "2 point 3". If the period is interpreted like a string as I intended, the 2's would be compared, then the periods, then the value "10" vs. the value "3", leading to the correct ordering of 2.3 followed by 2.10.

**Comment by daniel-beck** [comment]
@dmkoelle Interesting find; it could be considered to make sense to give decimal numbers special behavior (sorting by value). But then it stops being locale independent -- Germany has `1.234,56`; Switzerland uses both `1'234.56` and `1'234,56` -- and just generally becomes a mess. Especially when version numbers are mistaken as numbers…

**Comment by ssbarnea** [comment]
@dmkoelle You are right, current proposal is bugged due to the line you mentioned before. That's because my proposal used the "official" JS implementation from www.davekoelle.com/files/alphanum.js which contains this bug.

If you could create an alphanum repository on github where you keep all implementations it would be much easier to raise bugs or make pull requests for fixing them (or even better, to write some unittests).

I will update my change with the fix, which worked well in my tests. Also I should mention that there is already a published npm package that implements that. There are multiple npm packages published but that the only one that I tried, and it worked.

As I hate duplicating code across projects I would prefer finding a way to use external one but I am not aware of Jenkins being able to do this with npm dependencies. Also the licensing may be another issue preventing this. I would let others with more knowledge about the subject to decide on that matter.

**Comment by johnou** [comment]
@ssbarnea instead of modifying an existing javascript file and adding in custom code I think it would be less maintenance burden to simply update the entire sortable.js file which now supports alphanum sorting.. https://www.kryogenix.org/code/browser/sorttable/sorttable.js

**Comment by ssbarnea** [comment]
@johnou I totally agree with you and I realised that Jenkins is already using npm packages, so I made my first npm package an hour ago. The code is still buggy but at least it has CI https://github.com/jsContribs/alphanum 

Once I have the module working (regarding of using @dmkoelle implementation or another one), I will update this PR use a npm package.

**Comment by ssbarnea** [comment]
I could use some hints here. I managed to publish `sorttable` to npm and add it to `war/package.json` but I don't know how to replace the current loading of the embedded `sortable.js` from `core/src/main/resources/lib/layout/html.jelly` and `core/src/main/resources/lib/layout/layout.jelly` with the npm module.

**Comment by daniel-beck** [comment]
@kzantow Could you help out here? (Previous comment) Thanks!

**Comment by oleg-nenashev** [comment]
@kzantow @recena help with reviews will be appreciated.

So far there is a merge conflict after #3125 .

**Comment by oleg-nenashev** [comment]
I have resolved the merge conflict, but I suspect it is maximum what my Javascript skills allow here.
Is anybody interested in getting it over the line?

**Comment by daniel-beck** [comment]
The last notable activity here was almost two years ago, so I'm closing this PR as stalled. I'm sorry we were unable to support you with your NPM change.

FWIW I still think this would be viable and useful.

**Pull request closed without merging**